### PR TITLE
Refactor TropCyclone for external use in other modules

### PR DESCRIPTION
**Pull request by tovogt**
Changes proposed in this PR:
- The code for the TC wind computation is refactored with three main points in mind:
  1. The wind computation shouldn't use excessive amounts of memory even if the temporal or spatial resolution is increased.
  2. The code for the wind profiles should be reusable. I will use this in the forthcoming implementation of `TCRain` in climada_petals (see https://github.com/CLIMADA-project/climada_petals/pull/85).
  3. The wind computation should use SI units everywhere to make the physical equations more transparent. This is in accordance with the new `TCRain` code (see https://github.com/CLIMADA-project/climada_petals/pull/85).
- This also replaces all usage of `sparse.csr.csr_matrix` with `sparse.csr_matrix` to avoid warning messages.

### PR Author Checklist

- [x] Read the [Contribution Guide][contrib]
- [x] Correct target branch selected (if unsure, select `develop`)
- [x] Descriptive pull request title added
- [x] Source branch up-to-date with target branch
- [x] [Documentation](https://climada-python.readthedocs.io/en/latest/guide/Guide_PythonDos-n-Donts.html#2.--Commenting-&-Documenting) updated
- [x] [Tests][testing] updated
- [x] [Tests][testing] passing
- [x] No new [linter issues][linter]
- [x] [Changelog](https://github.com/CLIMADA-project/climada_python/blob/main/CHANGELOG.md) updated

### PR Reviewer Checklist

- [x] Read the [Contribution Guide][contrib]
- [x] [CLIMADA Reviewer Checklist](https://climada-python.readthedocs.io/en/latest/guide/Guide_Reviewer_Checklist.html) passed
- [x] [Tests][testing] passing
- [x] No new [linter issues][linter]

[contrib]: https://github.com/CLIMADA-project/climada_python/blob/main/CONTRIBUTING.md
[testing]: https://climada-python.readthedocs.io/en/latest/guide/Guide_Continuous_Integration_and_Testing.html
[linter]: https://climada-python.readthedocs.io/en/stable/guide/Guide_Continuous_Integration_and_Testing.html#3.C.--Static-Code-Analysis

### CODE REVIEW

**Comment by bguillod** [comment]
I very much like the changes. In particular tackling the memory and the physical units points is great and very much needed. Thanks a lot for the hard work @tovogt! From my first quick look it looks good to me, but I've not done a proper review. If you want me to, let me know and I'll see what I can do. If you have enough other people to review then I'm also ok not to dig deeper.

**Comment by tovogt** [comment]
@bguillod Thanks a lot for your feedback, Benoit! @ThomasRoosli will probably have a look, but your review would be very much appreciated if you can afford the time.

**Review by ThomasRoosli — COMMENTED** [review]
Review part 1, I hope I have more time early next week to go into the details of the compute_windfields-family of functions...
- Most files just contain the same change of a sparse command. The main changes in the PR are in trop_cyclone.py and test_trop_cyclone.py.
- In test_trop_cyclone.py a lot of the previous test (especially `test_windfield_models` remain unchanged, which ensures that the refactoring did not introduce a different behaviour.
- I advise to work on the docstrings of the functions that now accept a single track instead of several items). (see specific comment in the code).
- The main refactoring happens within the method `from_single_track` and its substructure
- I find the names `compute_windfield`, `_compute_windfields_sparse`, `_compute_windfields_sparse_chunked` and `_compute_windfields_chunked` confusingly similar. Especially since the they seems to call each other in a circular fashion. Maybe I have a suggestion after a more thorough review.
- As a general comment: I wonder if it would make sense to introduce a config entry for `max_memory_gb` so that the memory usage could be configured without changing code.
- As an even more general comment: I wonder if we should introduce a "peak_max_memory_gb" configuration value in the climada config, so that `max_memory_gb` could be derived from that and also other memory intensive functions could be configured the same way. This would of course only trigger an additional issue for now, and does not need to be implemented.

File: climada/hazard/trop_cyclone.py [hunk 1]
```diff
@@ -856,51 +1174,42 @@ def _close_centroids(
     # for each centroid, check whether it is in the buffer for any of the track positions
     return mask.any(axis=0)
 
-def _vtrans(
-    t_lat: np.ndarray,
-    t_lon: np.ndarray,
-    t_tstep: np.ndarray,
-    metric: str = "equirect"
-) -> Tuple[np.ndarray, np.ndarray]:
+def _vtrans(si_track: xr.Dataset, metric: str = "equirect"):
```
[the remark this answers is not in the document]
Comment (tovogt) [reply climada/hazard/trop_cyclone.py:1172]: Good idea, done. :)

File: climada/hazard/test/test_trop_cyclone.py [hunk 2]
```diff
@@ -46,6 +48,25 @@
...
+    def test_memory_limit(self):
+        """Test from_tracks when memory is (very) limited"""
+        tc_track = TCTracks.from_processed_ibtracs_csv(TEST_TRACK)
+        tc_track.equal_timestep()
+        tc_track.data = tc_track.data[:1]
+        # A very low memory constraint forces the algorithm to split the track into chunks.
+        # This should not affect the results. In practice, chunking is not applied due to limited
+        # memory, but due to very high spatial/temporal resolution of the centroids/tracks. We
+        # simulate this situation by artificially reducing the available memory.
+        tc_haz = TropCyclone.from_tracks(tc_track, centroids=CENTR_TEST_BRB, max_memory_gb=0.001)
```
[the remark this answers is not in the document]
Comment (tovogt) [reply climada/hazard/test/test_trop_cyclone.py:60]: I just took an arbitrary value that produces a number of chunks larger than 1. The number of chunks does only depend on the `max_memory_gb` parameter and on track/centroids input data. It is not environment-specific and won't change from system to system or with future changes to the environment.

**Review by ThomasRoosli — APPROVED** [review]
Thanks @tovogt for the adapted changes. 

- The restructuring of the `_compute_windfields...` family (and your explanation) helped me alot. Also with the new changes the loop into `_compute_windfields_sparse_chunked` is only done once (and it is more directly visible that it does so), which helps the overall understandability alot.
- The doc-strings of functions changeing the track-Dataset now contain very useful information about the elements used and added.
- I am happy with leaving the `max_memory_gb` as a parameter to the function without linking it to a global configuration variable
- In terms of code quality, understandability and tests and pylint I have nothing else to add.
Thanks for this contribution.

File: climada/hazard/trop_cyclone.py [hunk 3]
```diff
@@ -804,25 +975,162 @@ def compute_windfields(
...
+        ("central_pressure", "cen", "Pa"),
+        ("max_sustained_wind", "vmax", "m/s"),
+    ]
+    for long_name, var_name, si_unit in configs:
+        unit = track.attrs[f"{long_name}_unit"]
+        unit = unit_replace.get(unit, unit)
+        try:
+            conv_factor = ureg(unit).to(si_unit).magnitude
+        except:
+            raise ValueError(
```
[the remark this answers is not in the document]
Comment (tovogt) [reply climada/hazard/trop_cyclone.py:1042]: Good point, thanks for fixing it!

**Comment by tovogt** [comment]
Thanks for your review, @ThomasRoosli !

@emanuel-schmid The failing Jenkins "Post Actions" are nothing to worry about, right? It seems to me like they are not related to this PR.

**Comment by tovogt** [comment]
For testing, I computed all IBTrACS tracks for 1950-2021, at 300 arc-seconds and 0.5-hourly resolution, and with limited memory. I didn't truncate for high latitudes (the default setting for the wind field computations is truncation at 60° latitude). The chunking worked fine, except from Hurricane FAITH (1966):
image
This hurricane is quite a beast, traveling to more than 80 degrees latitude, so that I had to adapt the centroid pre-filtering step again in [a3d0f0e](https://github.com/CLIMADA-project/climada_python/pull/749/commits/a3d0f0ea3cee21e2d519e3b55765804311f30e86).

**Pull request merged**